### DISCUSSION TITLE
How to extend Typegoose document - generic class level

### DISCUSSION THREAD

**Discussion by RemyMachado** [typegoose/typegoose#703] [Q&A]
I was previously extending mongoose document to handle the dataSource model methods in a type-safe way and I'm stuck trying to reproduce it with typegoose & typegraphql.

*Mongoose:*
By extending mongoose.Document, I ensure that the type of the values I pass by are correct:
```typescript
import { MongoDataSource } from 'apollo-datasource-mongodb'
import { AnyKeys, Document } from 'mongoose'

class MongooseCrud<MongooseDocument extends Document> extends MongoDataSource<MongooseDocument> {
    async create(values: AnyKeys<MongooseDocument>) {
        return this.model.create(values)
    }
}
```

*still Mongoose:*
```typescript
const fooSchema = new Schema<IFoo>(
    {
        name: { type: String, required: true },
    },
    {
        collection: 'foos',
    }
)
    
await mongoose.connect(url)
const db = mongoose.connection.useDb(dbName)

const context = {
    foos: new MongooseCrud(db.model('Foo', fooSchema)),
}
```

*Typegoose:*
Now I'm trying with Typegoose, and I have struggle typing my context type with `getModelForClass`:
```typescript
type Context = {
    foos: MongooseCrud<any> // here I can't find the right type
}

@TypeGraphQL.ObjectType()
export class FooClass {
    @TypeGraphQL.Field((type) => TypeGraphQL.ID)
    id!: Scalars['ID']

    @TypeGraphQL.Field((type) => String)
    name!: Scalars['String']
}

const context = {
    foos: new MongooseCrud(getModelForClass(FooClass)), // trying to match this type
}
```

I probably also need to change this type, but I don't know what to use once again:
```typescript
class MongooseCrud<MongooseDocument extends Document> extends MongoDataSource<MongooseDocument> {

}
```

I'm trying to have typescript throwing errors if I give something else than `FooClass` for `context.foos`

I'll try to clarify if this is not clear enough.
Any help would be **greatly** appreciated!

**Comment by hasezoey** [comment]
> // trying to match this type

if you dont know which type to use, look at the type signature of the function the library gives back, in this case the type you are searching for is `ReturnModelType<YourTypegooseClassHere>`

---

as for the generic for `MongooseCrud`, i dont know what it is used for, so i cannot give a proper answer, but the most equivalent type for this would be the same, just wrap the input argument in a `DocumentType<YourTypegooseClassHere>` or if you decide to be only working with typegoose (no native mongoose), maybe consider replacing that generic with a generic for the class directly

**Reply by RemyMachado** [reply]
Thank you very much.
---
I feel like I misused your instructions, but it seems to work. I tried again, but I can't use `ReturnModelType<YourTypegooseClassHere>`. I'm getting the following error:
```
TS2344: Type 'ReturnModelType<FooClass, BeAnObject>' does not satisfy the constraint 'Document<any, any, any>'.   Type 'Model<any, BeAnObject, {}, {}> & FooClass' is missing the following properties from type 'Document<any, any, any>': $getAllSubdocs, $ignore, $isDefault, $isDeleted, and 34 more.
```
I didn't succeed to use it, but I made it work with DocumentType<FooClass> :

```typescript
type Context = {
     foos: MongooseCrud<DocumentType<FooClass>>
}
```

I kept this, and it seems to works fine with type-safe parameters :
```typescript
class MongooseCrud<MongooseDocument extends Document> extends MongoDataSource<MongooseDocument> {
    async create(values: AnyKeys<MongooseDocument>) {
        return this.model.create(values)
    }
}
```
---
> just wrap the input argument in a `DocumentType<YourTypegooseClassHere>`

I'm not sure to follow you. How do I make the method generic if I have to put an explicit class?

**Reply by hasezoey** [reply]
> I tried again, but I can't use ReturnModelType<YourTypegooseClassHere>. I'm getting the following error:

your original code was a generic with a *mongoose document*, but you were trying to get the type for a [*typegoose model*](https://typegoose.github.io/typegoose/docs/api/types/return-model-type/), which i had provided and you ended up using a [*typegoose document*](https://typegoose.github.io/typegoose/docs/api/types/document-type/)

> I'm not sure to follow you. How do I make the method generic if I have to put an explicit class?

depending on what you may want to do with your `Curd` class, you would end up with a generic argument over the *typegoose document* and *typegoose model*, to simplify this you could merge them to a single generic to use the *typegoose class* and change the places you would have `value: MongooseDocument` to `value: DocumentType<InputClass>`

the following is untested code, but should work similar enough:

```ts
class Crud<TClass extends AnyParamConstructor<any>> {
  somefunction(model: ReturnModelType<T>, document: DocumentType<T>) {}
}
```
